### PR TITLE
[PoC] Initial ideas for an Okio Store

### PR DESCRIPTION
**Pull request by bnorm**
Looking at FileOperator from OkHttp and potentially my own need for random access file operations, I've put together some ideas for an Okio native Store.

Still a bunch to do like documentation and multiplatform support, but I thought this would be a good place to discuss. Do you still want this abstraction in Okio? What should the API require things like `size()` and `truncate()`?

Feel free to close this pull request if this is not something you want in Okio. Otherwise, let me know what changes you would like to see and I will work toward getting this pull request ready for merge.

### CODE REVIEW

File: okio/src/jvmMain/kotlin/okio/Okio.kt [hunk 1]
```diff
@@ -108,6 +141,12 @@ private class InputStreamSource(
   override fun toString() = "source($input)"
 }
 
+fun FileChannel.store(): Store {
```
Comment (JakeWharton) [inline okio/src/jvmMain/kotlin/okio/Okio.kt]: Should be `asStore`.

"as" is for encapsulation, "to" is for transformation. We messed up with existing source and buffer functions.

Comment (bnorm) [reply okio/src/jvmMain/kotlin/okio/Okio.kt]: I agree. If a new convention is to be created - which it sounds like there is agreement - should the old convention be deprecated with a warning? I can create a separate pull-request for the existing functions.

**Comment by bnorm** [comment]
I used the name Store as suggested by #54. Naming is hard and I tend not to be very good at it so happy to change to whatever is suggested.

I definitely like the idea of separating out a read and write interface like Source and Sink. I'll go with the names Input and Output respectively unless something else is suggested. Some additional ideas:

- It might be helpful to have a union interface for things like `FileChannel` that can act as both (like Store is currently) but maybe the responsibility should be on the caller to create an Input and Output if they are both wanted.
- Like `Buffer`, there could be a Packet type which is a union of Input and Output and uses `Segments` to read and write data.
- `ByteString` could implement Output.

**Comment by swankjesse** [comment]
Suppose we assume SQLite is the best choice for its use case: mutable structured data stored in a file.

Is this true? I think it might be, and if so we should do things like make OkHttp’s overdue cookie storage depend on it.

If it's true what use cases are up for grabs here?

**Comment by swankjesse** [comment]
I hate files and filesystems. 

Every interaction with files is difficult:

 - It’s hard or impossible to do anything atomically. Partial success is the worst.
 - Nothing is portable. Filesystems have inconsistent semantics between Windows and UNIX. On Windows you can’t delete a file if somebody is reading it!
 - Permissions exist and must be managed.
 - Is it a local SSD? Or a spinning floppy disk on an NFS mount? No natural way to tell, but the consequences are real.
 - You can fill up the disk.

I think there’s only a few sane ways to do files:
 - In managed cloud databases like S3, DynamoDB, and TiDB.
 - In local sqlite databases
 - Doing a streaming write of `file.tmp` from beginning to end, atomically renaming `file.tmp` to `file` upon success
 - Reading any local file that is immutable

I genuinely think your `Store` API is great! It’s simple and exactly what I’d reach for if I needed to do random-access edits of a file. But I don’t think most developers should attempt to do random-access edits of files. It’s too difficult to get right, and there are too many failure modes.

Would you consider hosting this in a separate project? I don’t think this code is bad; I just don’t want to support it!

**Comment by bnorm** [comment]
As someone who had a job testing a NAS, I concur; file systems are very difficult.

I've got an [okiox](https://github.com/bnorm/okiox) project I started a long time ago which I've started to pick back up again recently to experiment with a coroutines based API. Maybe I'll add a `okiox-store` module to it and actually release it.

**Comment by swankjesse** [comment]
Awesome, yep that sounds good.

**Pull request closed without merging**